server/src/main/java/org/eclipse/openvsx/util/LogUtils.java

```diff
@@ -0,0 +1,827 @@
+   *      ss.updateAllReferences();
+   *      dataStore.addAll(ss.query(ts, trace));
+   *  }
+   * </pre>
+   *
+   * <p>will generate the following trace
+   *
+   * <pre>
+   * {@code trace output}
+   *  INFO: {"ts":12345,"ph":"B",tid:1,"name:Perform Query"}
+   *  INFO: {"ts":"12366,"ph":"E","tid":1}
+   * </pre>
+   */
+  public static class ScopeLog implements AutoCloseable {
+
+    private final long fTime;
+    private final long fThreadId;
+    private final Logger fLogger;
+    private final Level fLevel;
+    private final String fLabel;
+    private final Map<String, Object> fData = new HashMap<>();
+
+    /**
+     * Scope logger constructor
+     *
+     * @param log the JUL logger to log to
+     * @param level the log level see {@link Level}
+     * @param label The label of the event pair
+     * @param args Additional messages to pass for this scope, should be in pairs key, value, key2,
+     *     value2.... typically arguments. Note that these arguments will be logged only at the
+     *     beginning of the scope
+     */
+    public ScopeLog(Logger log, Level level, String label, Object... args) {
+      fTime = System.nanoTime();
+      fLogger = log;
+      fLevel = level;
+      fThreadId = Thread.currentThread().getId();
+      fLabel = label;
+      fLogger.log(
+          fLevel,
+          (() -> {
+            StringBuilder sb = new StringBuilder();
+            sb.append('{');
+            appendCommon(sb, 'B', fTime, fThreadId);
+            appendName(sb, fLabel);
+            appendArgs(sb, args);
+            sb.append('}');
+            sb.append('\n');
+            return sb.toString();
+          }));
+    }
+
+    /**
+     * Add a tag to the scope logger, will be written at the exit. This can save space on the trace
+     * by having a field appended to an event rather than writing a whole new event for a small
+     * chunk of data.
+     *
+     * <p>If the timing information is important than it would be more appropriate to call {@link
+     * LogUtils#traceInstant(Logger, Level, String, Object...)}
+     *
+     * @param name the name of the field
+     * @param value The value of the field.
+     */
+    public void addData(String name, Object value) {
+      fData.put(name, value);
+    }
+
+    @Override
+    public void close() {
+      long time = System.nanoTime();
+      fLogger.log(
+          fLevel,
+          (() -> {
+            StringBuilder sb = new StringBuilder();
+            sb.append('{');
+            appendCommon(sb, 'E', time, fThreadId);
+            return appendArgs(sb, fData).append('}').append('\n').toString();
+          }));
+    }
+  }
+
+  /**
+   * Builder class for the {@link FlowScopeLog}. One can either set a category or a parent scope
+   * before building the flow scope log. If none is set, a default category called "null" will be
+   * used.
+   *
+   * @author Geneviève Bastien
+   */
+  public static class FlowScopeLogBuilder {
```

Comment:
I can't find any usages of the class in the other parts of the code. It is really necessary there?

Reply:
For future use, we use it in trace compass to have spans with links.